### ISSUE TITLE
Download counts are going backwards again

### ISSUE THREAD

**Issue by rianjs** [NuGet/NuGetGallery#7053]
I've opened several tickets in the past about download counts doing weird things. Today the download count went backward again.

Fri, 5-Apr-2019 | 579,211 | 3,925
-- | -- | -- 
Sat, 6-Apr-2019 | 573,753 | -5,458

I suspect stats have been slightly off for a while, because I see consistently high download counts on weekends, when historically those days have been much, much lower. It almost feels like the stats jobs are "catching up" on the weekends with stats they didn't count during the week. I didn't report it, because I didn't have anything concrete to say aside from "This seems off."

Today it's concrete: download counts went backwards.

**Comment by joelverhagen** [comment]
Thanks for taking the time to report this!

Can you provide exact steps on where you are seeing the download counts? Exact URLs and timestamps (as close as possible) would help us diagnose the issue. Screenshots would help too.

Also how long did the lower number "stick"? Did you see it jump back to the higher number (or a number even higher than before)? If so, how long did it take?

**Comment by rianjs** [comment]
Every morning around 6:15 Eastern US time I check the total download count of my package

https://www.nuget.org/packages/Ical.Net/

This morning it jumped back into positive territory at 582,421.

**Comment by rianjs** [comment]
This is the link to my Google spreadsheet. It shows the regression and then the big forward jump. And Sunday’s totals are too low to be accurate based on past history.

https://docs.google.com/spreadsheets/d/1aPisoc9KQDlggyfuFjo8AePuKkQD5tVonZpdmCzU1SE

**Comment by xavierdecoster** [comment]
@joelverhagen No idea, only changes to this pipeline in past months are the ones made by @agr. (which doesn't necessarily mean those changes are related to whatever is happening)

@cristinamanum @zhhyu do we know whether this happens due to the 6-weeks roll-up?

**Comment by zhhyu** [comment]
I also notice that in the NuGet statistic page, the oldest week's download accounts are different among different VMs. (Try to refresh the page for several times). If it's due to the inconsistency of caching on different VMs, why is the oldest week's data not same?

**Comment by cristinamanum** [comment]
It is hard to say if the issue is because of the RollUpDownloadFacts. The job has more failed transaction than usual but this does not mean that this is the root cause for the behavior.

**Comment by cristinamanum** [comment]
Hardened the stats injection to prevent degraded counts to be ingested.
Add additional telemetry to help diagnose this type of issues in the future.

**Comment by cristinamanum** [comment]
Reopening as the issue was manifested again. https://github.com/NuGet/NuGetGallery/issues/7258

Previous [job ](https://github.com/NuGet/NuGet.Jobs/pull/745) update was on the Stats.AggregateCdnDownloadsInGallery that aggregates the data from Stats db and ingests it in the gallery db. 

However the [ObjectMaterializedInterception](https://github.com/NuGet/NuGetGallery/blob/1c11e12266c97d57a4bc0bec022c001fdba3dc14/src/NuGetGallery.Core/Entities/Interception/ObjectMaterializedInterception.cs#L8)
 hijacks the data read from the DB and overwrite with the data from downloads.v1.json blob that is updated by the job Stats.CreateAzureCdnWarehouseReports. 
From the initial investigation:
1.  Changes similar with https://github.com/NuGet/NuGet.Jobs/pull/745 will need to be applied to teh job. 
2. The logic of blob refresh in Gallery is done through a task scheduled at each 15 minutes. https://github.com/NuGet/NuGetGallery/blob/1c11e12266c97d57a4bc0bec022c001fdba3dc14/src/NuGetGallery/App_Start/AppActivator.cs#L279. If this task fail to refresh on a instance of the gallery that instance will show stale data . 

- More telemetry may be added on this route. 
- A force refresh may be also be implemented based on timestamp built in the json file. 
- Error for blob refresh are added to elmah, it would be great to be added to AI and maybe custom metrics for the failures count as this impacts user experience. https://github.com/NuGet/NuGetGallery/blob/1c11e12266c97d57a4bc0bec022c001fdba3dc14/src/NuGetGallery/App_Start/AppActivator.cs#L290.

**Comment by cristinamanum** [comment]
https://github.com/NuGet/NuGetGallery/issues/7258

**Comment by ryuyu** [comment]
So we investigate this a bit and believe we have located the root of the issue.

We have a job that runs every 12 hours that tries to compact any download records outside of 6 weeks into a single record (this work is done by the [Stats.RollUpDownloadFacts job](https://github.com/NuGet/NuGet.Jobs/tree/master/src/Stats.RollUpDownloadFacts) using this [stored procedure](https://github.com/NuGet/NuGet.Jobs/blob/master/src/Stats.Warehouse/Programmability/Stored%20Procedures/dbo.RollUpDownloadFacts.sql))

We also have a job that runs every 3 hours and generates the report that is used to display statistics in the gallery (this work is done by the [Stats.CreateAzureCdnWarehouseReports](https://github.com/NuGet/NuGet.Jobs/tree/master/src/Stats.CreateAzureCdnWarehouseReports) using this [stored procedure](https://github.com/NuGet/NuGet.Jobs/blob/master/src/Stats.Warehouse/Programmability/Stored%20Procedures/dbo.SelectTotalDownloadCountsPerPackageVersion.sql))

If you examine https://github.com/NuGet/NuGet.Jobs/blob/master/src/Stats.Warehouse/Programmability/Stored%20Procedures/dbo.SelectTotalDownloadCountsPerPackageVersion.sql, you will see that the (NOLOCK) query hint is passed in a lot of places. This was originally so that we could get up to date statistics, and not have the job that generates the reports get blocked by the job that was trying to roll up the stats.

We believe that NOLOCK query hint is causing the second job to see dirty rows that the RollUp job is create/deleting as part of its transactions. We noticed that there seem to be some package id/versions that "lose" downloads when the reports are generated while the RollUp job is running. If we compared 2 reports generated while the RollUp job was NOT running, there were no instances of download counts decreasing.

We are still investigating a fix for this. We have explored using a replica to run the report generation, but this had the same issue. We are currently exploring different query hints that will allow us to read without blocking and with some data consistency, but are running into some performance issues.

We will continue to look into this and update as we have more information.
Thank you for your patience.

**Comment by scottbommarito** [comment]
Looked into this some more. As said by @ryuyu, it appears to be the `NOLOCK` that is causing the dirty reads. [`NOLOCK` is the equivalent of setting the transaction isolation to `READ UNCOMMITTED` for a single table](https://stackoverflow.com/questions/27198020/what-is-are-difference-between-nolock-and-uncommitted), which is obviously not what we want for this--we want our download reports to be generated with finalized data.

What we want is a transaction isolation level that both does not lock and reads fully committed data. Ironically, this is the default behavior in SQL Azure without `NOLOCK`. SQL Server's default isolation of all transactions is `READ COMMITTED`, which has two implementations. If the `ALLOW_SNAPSHOT_ISOLATION` and `READ_COMMITTED_SNAPSHOT` settings are on for a database, the `READ COMMITTED` reads from a snapshot of the table that is not dirty and does not lock. If either setting is off, it uses a locking implementation. SQL Azure defaults both of these settings to on, and I confirmed this on our DEV databases. [Here's some general documentation on snapshot isolation.](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql/snapshot-isolation-in-sql-server)

If we remove the `NOLOCK`s from the `SelectTotalDownloadCountsPerPackageVersion` stored procedure mentioned above, we should fix this behavior. On that note, we should reanalyze how we use `NOLOCK` in all our stored procedures, because it is likely being used improperly in them as well.

**Comment by joelverhagen** [comment]
We have since moved to a new statistics pipeline. This issue should not happen anymore. If you run into it again, please feel free to open a new issue. Thanks!
